File: leetcode/longest_increasing_subsequence.go
// https://leetcode.com/problems/longest-increasing-subsequence/solution/
// 300. Longest Increasing Subsequence
package main

import "fmt"

func lengthOfLIS(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	tab := make([]int, len(nums))
	tab[0] = 1
	res := 1
	for i := 1; i < len(nums); i++ {
		max := 0
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] {
				if max < tab[j] {
					max = tab[j]
				}
			}
		}
		tab[i] = max + 1
		if res < tab[i] {
			res = tab[i]
		}
	}
	return res
}

func main() {
	nums := []int{1, 3, 5, 6, 7, 4, 6, 7, 8, 1, 4, 5}
	res := lengthOfLIS(nums)
	fmt.Println(res)
}

File: leetcode/longest_increasing_subsequence.py
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        memo = [1] * len(nums)
        m = 1
        for i in reversed(range(0, len(nums))):
            for j in range(i, len(nums)):
                if nums[i] < nums[j]:
                    memo[i] = max(memo[i], (memo[j]) + 1)
                    m = max(memo[i], m)

        return m


if __name__ == '__main__':
    nums = [1, 2, 3, 4, 5, 6, 4, 3, 4, 3, 2, 12]
    res = Solution().lengthOfLIS(nums)
    print(res)
